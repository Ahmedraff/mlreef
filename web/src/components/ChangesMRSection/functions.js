import { Base64 } from 'js-base64';
import Base64ToArrayBuffer from 'base64-arraybuffer';
import { getFileDifferences } from '../../functions/apiCalls';
import CommitsApi from 'apis/CommitsApi';
import { isImageFormat } from 'functions/dataParserHelpers';

const commitsApi = new CommitsApi();

export const fetchFileDiffInfo = (projectId, aheadCommits) => Promise.all(aheadCommits.map(async (commit, index) => {
  const { id: commitId } = commit;
  const commitDiffs = await commitsApi.getCommitDiff(projectId, commitId, 0, false);
  return await Promise.all(commitDiffs?.body?.map(async (commitDiff) => {
    const {
      previousVersionFile,
      nextVersionFile,
      imageFileSize,
    } = await getFileDifferences(projectId, commitDiff, commit.parent_ids[0], commit.id);
    let previousVersionFileParsed = previousVersionFile;
    let nextVersionFileParsed = nextVersionFile;
    let sizeDeleted = 0;
    let sizeAdded = 0;
    const fileName = commitDiff.old_path.split('/').slice(-1)[0];
    if (previousVersionFile && !isImageFormat(fileName)) {
      previousVersionFileParsed = Base64.decode(
        Base64ToArrayBuffer.encode(previousVersionFile),
      );
      sizeDeleted = previousVersionFile.byteLength;
    }
    if (nextVersionFile && !isImageFormat(fileName)) {
      nextVersionFileParsed = Base64.decode(Base64ToArrayBuffer.encode(nextVersionFile));
      sizeAdded = nextVersionFile.byteLength;
    }
    const path = commitDiff.new_path
      ? commitDiff.new_path
      : commitDiff.old_path;
    const fileExtension = path.split('.').pop().toLowerCase();
    return {
      fileName,
      fileExtension,
      previousVersionFileParsed,
      nextVersionFileParsed,
      id: `${commit.id} ${index}`,
      sizeDeleted,
      sizeAdded,
      imageFileSize,
    };
  }));
}));
