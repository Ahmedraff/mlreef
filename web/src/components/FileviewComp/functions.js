import FilesApi from 'apis/FilesApi.ts';
import CommitsApi from 'apis/CommitsApi.ts';

const commitsApi = new CommitsApi();

const filesApi = new FilesApi();

const getCommit = (projectId, lastCommitId) => commitsApi
  .getCommitDetails(projectId, lastCommitId);

const getFileAndInformation = (gid, file, branch, commit) => filesApi
  .getFileData(gid, file?.includes('/') ? encodeURIComponent(file) : file, branch || commit)
  .then(async (fData) => {
    const commitInfoDet = await getCommit(gid, fData.last_commit_id);
    return {
      fData,
      commitInfoDet,
    };
  });

export default {
  getFileAndInformation,
};
