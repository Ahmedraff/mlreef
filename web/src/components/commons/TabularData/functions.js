import genUUID from 'uuid/v1';

const parseNumber = (text) => {
  const num = parseFloat(text);

  return Number.isNaN(num) ? text : num;
};

export const transpose = (matrix) => matrix[0].map((_, ci) => matrix.map((row) => row[ci]));

const sanatizeMatrix = (matrix) => {
  const head = matrix[0];
  const tail = matrix[matrix.length - 1];

  const sameLength = head.length === tail.length;
  const lastRowfilled = tail.every((col) => col !== undefined);

  return sameLength && lastRowfilled ? matrix : matrix.slice(0, -1);
};

export const cvsToArray = (text, separator = ',') => {
  const matrix = text.split('\n')
    .map((line) => line.split(separator)
      .map(parseNumber));

  return sanatizeMatrix(matrix);
};

export const jsonToArray = (input) => {
  const items = Array.isArray(input) ? input : Object.values(input);

  if (!items.length) return items;
  // const row = Object.values(input)[0];
  const row = items[0];
  const header = Object.keys(row).map((label) => label);

  // const data = Object.values(input)
  const data = items
    .map((cols) => Object.values(cols).map((val) => val));

  const matrix = [header].concat(data);

  return sanatizeMatrix(matrix);
};

export const arrayToRichData = (matrix, separator = ',') => {
  const header = matrix[0];

  return {
    id: genUUID(),
    meta: {
      separator,
      timestamp: new Date(),
    },
    matrix,
    // data: flat(matrix.map((row, y) => row.map((value, x) => ({ x, y, value }))))
    data: matrix.map((cols, y) => ({
      id: y,
      cols: cols.map((value, x) => ({
        x,
        y,
        value,
        label: header[x],
      })),
    })),
  };
};
