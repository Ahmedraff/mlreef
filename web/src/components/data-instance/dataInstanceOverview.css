/* ------------- Card content ---------------*/

.card-content > .summary-data > .buttons-div > button{
    margin-right: 5px !important;
    margin-top: 28px;
}

/* ----------------------- Chart container ----------------------- */

.card-content > .data-summary {
    display: flex;
}

.card-content > .data-summary > .chart-container {
    width: 70%;
}

.card-content > .data-summary > .content {
    margin-left: 3%;
}

.card-content > .data-summary > .content > p { 
    margin: 4px;
}

/* Instance  dropdown*/

.save-instance {
    box-shadow: 0 2px 4px rgb(0, 0, 0, 0.3);
    position: absolute;
    display: block !important;
    width: 300px;
    height: auto;
    background: #fff;
    top: 50%;
    right: 0.5em;
    z-index: 1;
    border-radius: 4px;
}

.save-instance .search-branch > div:hover{
    background-color: #ececec;
    cursor: pointer;
}

#data-ins-summary-data {
  width: 100%;
}
