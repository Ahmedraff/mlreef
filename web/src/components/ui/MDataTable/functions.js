const extractField = (row, x) => row.cols.find((col) => col.x === x)?.value;

export const sortDesc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(b, x) - extractField(a, x));

export const sortAsc = (x) => (rows) => rows.slice()
  .sort((a, b) => extractField(a, x) - extractField(b, x));

export const round = (decs) => (input) => {
  const decimals = parseInt(decs, 10);

  if (Number.isNaN(decimals)) return input;

  if (decimals > 12) return input;

  const num = parseFloat(input);
  if (Number.isNaN(num)) return input;

  return Math.round(num * (10 ** decimals)) / (10 ** decimals);
};
