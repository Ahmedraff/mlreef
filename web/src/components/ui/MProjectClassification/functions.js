import { ALGORITHM, DATA_OPERATION, MODEL, OPERATION, VISUALIZATION } from "dataTypes";

export const getSearchable = (feSearchableTypes) => {
  switch (feSearchableTypes) {
    case DATA_OPERATION:
      return OPERATION;
    case MODEL:
      return ALGORITHM;
    default:
      return VISUALIZATION;
  }
}