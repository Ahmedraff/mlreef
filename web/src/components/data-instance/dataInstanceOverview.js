import React, { Component } from "react";
import Navbar from "../navbar/navbar";
import ProjectContainer from "../projectContainer";
import "./dataInstanceOverview.css";
import { connect } from "react-redux";
import filesApi from "../../apis/FilesApi";
import arrow_down_white_01 from "../../images/arrow_down_white_01.svg";
import { Instruction } from '../instruction/instruction';
import { getTimeCreatedAgo } from '../../functions/utilities';

const Active = "Active";
const Expired = "Expired";

class InstanceCard extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            showChart: false,
        };
    }

    handleButtonsClick(e) {
        const branchName = encodeURIComponent(e.currentTarget.parentNode.parentNode.getAttribute("data-key"));
        const pId = this.props.params.experiments[0].projId;
        this.props.history.push(`/my-projects/${pId}/master/data-instances/${branchName}`)
    }

    handleEmptyClick(e) {
        return;
    }

    getButtonsDiv(experimentState) {
        let buttons;
        switch (experimentState) {
            case Active:
                buttons = [
                    <button className="non-active-black-border experiment-button"
                        onClick={(e) => this.handleEmptyClick(e)}>
                        View Pipeline
                    </button>,
                    <button className="dangerous-red"><b>X</b></button>,
                    <Dropdown />
                ];
                break;
            case Expired:
                buttons = [
                    <button className="non-active-black-border experiment-button"
                        onClick={(e) => this.handleEmptyClick(e)}>
                        View Pipeline
                    </button>];
                break;
            default:
                break;
        }

        return (<div className="buttons-div">{buttons}</div>)
    }

    render() {
        const params = this.props.params;
        return (
            <div className="experiment-card">
                <div className="header">
                    <div className="title-div">
                        <p><b>{params.currentState}</b></p>
                    </div>
                </div>

                {params.experiments.map((experiment, index) => {
                    let modelDiv = "inherit";
                    let progressVisibility = "inherit";
                    if (experiment.currentState === "Expired")
                        progressVisibility = "hidden"
                    return (
                        <div key={index} className="card-content">
                            <div className="summary-data" data-key={`${experiment.descTitle}`}>
                                <div className="project-desc-experiment">
                                    <p onClick={(e) => {
                                        experiment.currentState === "Expired"
                                            ? this.handleEmptyClick()
                                            : this.handleButtonsClick(e)
                                    }}
                                        style={{ cursor: "pointer" }}>
                                        <b>{experiment.descTitle}</b>
                                    </p>
                                    <p>Created by <b>{experiment.userName}</b><br />
                                        {experiment.timeCreatedAgo} ago
                                    </p>
                                </div>
                                <div className="project-desc-experiment" style={{ visibility: progressVisibility }}>
                                    <p><b>Use: 24GB</b></p>
                                    <p>Expires in:{experiment.expiration}</p>
                                </div>
                                <div className="project-desc-experiment" style={{ visibility: modelDiv }}>
                                    <p><b>24,051 files changed</b></p>
                                    <p>Id: {experiment.di_id ? experiment.di_id : "72fb5m"}</p>
                                </div>
                                {this.getButtonsDiv(experiment.currentState)}
                            </div>
                        </div>)
                })
                }
            </div>
        )
    }
}


class DataInstanceOverview extends Component {
    constructor(props) {
        super(props);
        const project = this.props.projects.selectedProject;

        this.state = {
            project: project,
            branches: []
        };
    }

    componentDidMount() {
        filesApi.getBranches("gitlab.com", this.state.project.id)
            .then(res => res.json())
            .then(response =>
                this.setState({ branches: response.filter(branch => branch.name.startsWith("data-pipeline")) })
            )
    }

    handleButtonsClick(e) {
        e.target.parentNode.childNodes.forEach(childNode => {
            if (childNode.id !== e.target.id) {
                childNode.classList.remove("active-border-light-blue");
                childNode.classList.add("non-active-black-border");
            }
        });
        e.target.classList.add("active-border-light-blue");
        e.target.classList.remove("non-active-black-border");
    }

    hideInstruction = () =>
        document.getElementById("instruction-pipe-line").classList.add("invisible");

    render() {
        const project = this.state.project;
        return (
            <div id="experiments-overview-container">
                <Navbar />
                <ProjectContainer project={project}
                    activeFeature="data"
                    folders={['Group Name', project.name, 'Data', 'Instances']}
                />
                <Instruction
                    titleText={"Handling Data instances:"}
                    paragraph={
                        `A data instance is a reslt of an executed data pipeline, thus being a pre-processed data set. You can use these data sets
                        directly for training or to merge them into a data repository in order to permanently save the changes made.`
                    }
                />
                <div className="main-content">
                    <br />
                    <div id="line" />
                    <br />
                    <div id="buttons-container">
                        <button id="all" className="non-active-black-border experiment-button"
                            onClick={(e) => this.handleButtonsClick(e)}>
                            All
                        </button>
                        <button id="InProgress" className="non-active-black-border experiment-button"
                            onClick={(e) => this.handleButtonsClick(e)}>
                            In Progress
                        </button>
                        <button id="Active" className="non-active-black-border experiment-button"
                            onClick={(e) => this.handleButtonsClick(e)}>
                            Active
                        </button>
                        <button id="expired" className="non-active-black-border experiment-button"
                            onClick={(e) => this.handleButtonsClick(e)}>Expired
                        </button>
                    </div>

                    <InstanceCard params={
                        {
                            "currentState": Expired,
                            "experiments": [
                                {
                                    "currentState": Expired,
                                    "descTitle": "UNET_SARData",
                                    "userName": "Camillo Pachmann",
                                }
                            ]
                        }
                    }
                    />
                    {this.state.branches.map((item, index) => {
                        const timediff = getTimeCreatedAgo(item.commit.created_at);
                        return (
                            <InstanceCard key={index} history={this.props.history} params={
                                {
                                    "currentState": Active,
                                    "experiments": [
                                        {
                                            "currentState": Active,
                                            "di_id": item.commit.short_id,
                                            "descTitle": item.name,
                                            "userName": item.commit.author_name,
                                            "timeCreatedAgo": timediff,
                                            "expiration": timediff,
                                            "projId": this.state.project.id,
                                            "modelTitle": "Resnet-50",
                                        }
                                    ]
                                }
                            }
                            />
                        )
                    })}
                </div>
                <br />
                <br />
            </div>
        )
    }
}

function Dropdown() {
    const [state, setState] = React.useState(false);
    const node = React.useRef();

    const handleClickOutside = e => {
        if (node.current.contains(e.target)) {
            return;
        }
        setState(false);
    };

    React.useEffect(() => {
        if (state) {
            document.addEventListener("mousedown", handleClickOutside);
        } else {
            document.removeEventListener("mousedown", handleClickOutside);
        }
        return () => {
            document.removeEventListener("mousedown", handleClickOutside);
        };
    }, [state]);

    return (
        <>
            <button onClick={e => setState(!state)} ref={node} className="light-green-button experiment-button non-active-black-border">
                <span><b style={{ margin: "0 10px 10px 0" }}>Save</b></span>
                <img className="dropdown-white" src={arrow_down_white_01} alt="" />
            </button>
            {state &&
                <div className="save-instance">
                    <div
                        style={{ marginLeft: "25%", fontSize: "14px" }}>
                        <p>Save Data Instances</p>
                    </div>
                    <hr />
                    <div className="search-branch">
                        <div>
                            <p><b>New branch</b></p>
                            <p className="dull">Only new data is saved in new branch</p>
                        </div>
                        <div>
                            <p><b>Create Pull Request</b></p>
                            <p className="dull">New data and original data coexist in existing branch</p>
                        </div>
                    </div>
                </div>
            }
        </>
    )
}

function mapStateToProps(state) {
    return {
        projects: state.projects
    };
}

export default connect(mapStateToProps)(DataInstanceOverview);
