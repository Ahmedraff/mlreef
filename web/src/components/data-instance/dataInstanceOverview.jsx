import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { toastr } from 'react-redux-toastr';
import { shape, objectOf, func } from 'prop-types';
import DataPipelineApi from 'apis/DataPipelineApi';
import BranchesApi from 'apis/BranchesApi';
import Navbar from '../navbar/navbar';
import ProjectContainer from '../projectContainer';
import './dataInstanceOverview.css';
import Instruction from '../instruction/instruction';
import { getTimeCreatedAgo, parseToCamelCase } from '../../functions/dataParserHelpers';
import DataInstancesDeleteModal from '../data-instances-delete-and-abort-modal/dataInstancesDeleteNAbortModal';
import pipelinesApi from '../../apis/PipelinesApi';
import {
  RUNNING,
  SUCCESS,
  CANCELED,
  FAILED,
  PENDING,
} from '../../dataTypes';
import { deleteBranch, getBranchesList } from '../../actions/branchesActions';
import { fireModal } from '../../actions/actionModalActions';
import { setPreconfiguredOPerations } from '../../actions/userActions';
import { classifyPipeLines } from '../../functions/pipeLinesHelpers';

const getStatusForDataInstance = (status) => {
  let mappedStatus = status;
  switch (status) {
    case RUNNING:
      mappedStatus = 'In progress';
      break;
    case SUCCESS:
      mappedStatus = 'Active';
      break;
    case CANCELED:
      mappedStatus = 'Aborted';
      break;
    case PENDING:
      mappedStatus = 'In progress';
      break;
    default:
      break;
  }

  return mappedStatus;
};

export const InstanceCard = ({ ...props }) => {
  const {
    params,
    history,
    name,
    deletePipeline,
    setPreconfiguredOPerations,
  } = props;

  function goToPipelineView(id) {
    const pId = params.instances[0].projId;
    const instance = params.instances.filter((ins) => ins.id === id)[0];
    const configuredOperations = {
      dataOperatorsExecuted: instance.dataOperations,
      inputFiles: instance.inputFiles,
      pipelineBackendId: instance.pipelineBackendId,
    };
    setPreconfiguredOPerations(configuredOperations);
    history.push(`/my-projects/${pId}/pipeline-execution/new-data-pipeline`);
  }

  // returns a function that receives the name (or id) of the pipeline and deletes it
  function handleDelete() {
    const content = (
      <div style={{ textAlign: 'center' }}>
        <p>
          <strong>Are you sure you want to delete your dataset?</strong>
        </p>
        <p>
          You will free
          <strong>23Gb</strong>
          {' '}
          of cloud storage.
        </p>
        <p>You will loose all files generated by the pipeline.</p>
      </div>
    );

    const onPositive = () => deletePipeline(name);

    fireModal({
      title: `Delete dataset from your data pipeline ${name}`,
      type: 'danger',
      content,
      onPositive,
    });
  }

  function getButtonsDiv(instanceStatus, id) {
    let buttons;
    if (instanceStatus === RUNNING || instanceStatus === PENDING) {
      buttons = [
        <button
          type="button"
          key="abort-button"
          className="btn btn-danger border-solid my-auto"
          style={{ width: 'max-content' }}
        >
          Abort
        </button>,
      ];
    } else if (
      instanceStatus === SUCCESS
    ) {
      buttons = [
        <button
          type="button"
          key="experiment-button"
          className="btn btn-outline-dark my-auto"
          onClick={() => goToPipelineView(id)}
        >
          View Pipeline
        </button>,
        <button
          type="button"
          key="delete-button"
          onClick={
            () => props.setIsDeleteModalVisible(true, 'delete')
          }
          className="btn btn-danger btn-icon my-auto"
        >
          <i className="fa fa-times" />
        </button>,
      ];
    } else if (instanceStatus === FAILED
      || instanceStatus === CANCELED) {
      buttons = [
        <button
          type="button"
          key="experiment-button"
          className="btn btn-outline-dark my-auto"
          onClick={() => goToPipelineView(id)}
        >
          View Pipeline
        </button>,
        <button
          type="button"
          key="delete-button"
          className="btn btn-danger btn-icon my-auto"
          onClick={handleDelete}
        >
          <i className="fa fa-times" />
        </button>,
      ];
    }

    return (
      <div className="buttons-div d-flex">{buttons}</div>
    );
  }

  return (
    <div className="pipeline-card">
      <div className="header">
        <div className="title-div">
          <p><b>{getStatusForDataInstance(params.currentState)}</b></p>
        </div>
      </div>

      {params.instances.map((instance, index) => {
        const dataId = instance.id;
        const { timeCreatedAgo } = instance;
        const dataInstanceName = instance.descTitle;
        const uniqueName = dataInstanceName.split('/')[1];
        const modelDiv = 'inherit';
        let progressVisibility = 'inherit';
        if (instance.currentState === 'Expired') { progressVisibility = 'hidden'; }
        return (
          <div key={index} className="card-content">
            <div id="data-ins-summary-data" className="summary-data" data-key={`${instance.descTitle}`}>
              <div className="project-desc-experiment">
                <Link
                  to={{
                    pathname: `/my-projects/${instance.projId}/master/-/datasets/${dataId}`,
                    state: { di_name: encodeURIComponent(dataInstanceName), timeCreatedAgo },
                  }}
                >
                  <b>{uniqueName}</b>
                </Link>
                <p>
                  Created by
                  {' '}
                  <a href={`/${instance.userName}`}>
                    <b>
                      {instance.userName}
                    </b>
                  </a>
                  <br />
                  {instance.timeCreatedAgo}
                  {' '}
                  ago
                </p>
              </div>
              <div className="project-desc-experiment" style={{ visibility: progressVisibility }}>
                <p><b>Usage: ---</b></p>
              </div>
              <div className="project-desc-experiment" style={{ visibility: modelDiv }}>
                <p>
                  Id:
                  {' '}
                  {instance.id}
                </p>
              </div>
              { getButtonsDiv(instance.currentState.toUpperCase(), instance.id) }
            </div>
          </div>
        );
      })}
    </div>
  );
};

class DataInstanceOverview extends Component {
  constructor(props) {
    super(props);
    this.state = {
      project: {},
      all: [],
      isDeleteModalVisible: false,
      dataInstances: [],
      typeOfMessage: '',
      onPositiveAction: (x) => x,
    };

    this.fetchPipelines = this.fetchPipelines.bind(this);
    this.setIsDeleteModalVisible = this.setIsDeleteModalVisible.bind(this);
    this.deletePipeline = this.deletePipeline.bind(this);
  }

  componentDidMount() {
    const { getBranches, projects: { selectedProject: { gid } } } = this.props;
    getBranches(gid);
    this.fetchPipelines();
  }

  // this is a way not the optimal to pass a callback function to modal
  setIsDeleteModalVisible(isDeleteModalVisible, typeOfMessage = '', payload = {}) {
    const { entity, name } = payload;
    const onPositiveAction = entity === 'pipeline'
      ? this.deletePipeline(name)
      : (x) => x;

    this.setState({
      isDeleteModalVisible,
      typeOfMessage,
      onPositiveAction,
    });
  }

  fetchPipelines() {
    let project;
    let filteredbranches = [];
    let gid;

    const brApi = new BranchesApi();
    const { projects, projects: { selectedProject: { id } } } = this.props;
    const dataPipelineApi = new DataPipelineApi();

    if (projects) {
      project = projects.selectedProject;
      gid = project.gid;
    }

    dataPipelineApi.getProjectPipelines(id)
      .then((backendPipelines) => {
        const dataPipelines = backendPipelines.filter((pipe) => pipe.pipeline_type === 'DATA');
        brApi.getBranches(gid).then((branches) => {
          filteredbranches = branches.filter((branch) => branch.name.startsWith('data-pipeline'));
          pipelinesApi.getPipesByProjectId(gid)
            .then((res) => {
              const dataInstancesClassified = classifyPipeLines(res, filteredbranches, dataPipelines);
              this.setState({
                project,
                dataInstances: dataInstancesClassified,
                all: dataInstancesClassified,
              });
            });
        });
      })
      .catch(() => toastr.error('Error', 'Error getting the pipelines'));
  }

  deletePipeline(name) {
    // lookup the pipeline in every group's values
    const { project: { gid } } = this.state;
    const { deleteBranch, getBranches } = this.props;

    deleteBranch(gid, name)
      .then(() => getBranches(gid))
      .then(this.fetchPipelines)
      .catch((err) => {
        const message = (err && err.statusText) || err;
        toastr.error('Error', `Unable to delete: ${message}`);
      });
  }

  handleButtonsClick(e) {
    if (e) {
      e.target.parentNode.childNodes.forEach((childNode) => {
        if (childNode.id !== e.target.id) {
          childNode.classList.remove('active');
        }
      });
      e.target.classList.add('active');

      const { all } = this.state;
      let filteredInstances = all;
      if (e.target.id === 'InProgress') {
        filteredInstances = all.filter((exp) => exp.status === 'running');
      } else if (e.target.id === 'Active') {
        filteredInstances = all.filter((exp) => exp.status === 'success');
      } else if (e.target.id === 'expired') {
        filteredInstances = all.filter((exp) => exp.status === 'failed');
      }
      this.setState({ dataInstances: filteredInstances });
    }
  }

  render() {
    const {
      project,
      dataInstances,
      isDeleteModalVisible,
      typeOfMessage,
      onPositiveAction,
    } = this.state;
    const {
      history, projects: { selectedProject }, fireModal, setPreconfiguredOPerations,
    } = this.props;
    let groupName;
    let name;
    if (selectedProject) {
      groupName = selectedProject.namespace.name;
      name = selectedProject.name;
    }
    return (
      <>
        <DataInstancesDeleteModal
          isModalVisible={isDeleteModalVisible}
          setIsVisible={this.setIsDeleteModalVisible}
          typeOfMessage={typeOfMessage}
          onPositiveAction={onPositiveAction}
        />
        <div>
          <Navbar />
          { project.gid && (
            <ProjectContainer
              activeFeature="data"
              folders={[groupName, name, 'Data', 'Instances']}
            />
          )}
          <Instruction
            id="DataInstanceOverview"
            titleText="Handling datasets:"
            paragraph={
              `A dataset is the result of an executed data pipeline. You can use this dataset directly as your source of data for an experiment
               (or another data pipeline). You can also merge the dataset to your master branch in your repository.`
            }
          />
          <div className="main-content">
            <div id="buttons-container" className="left">
              <button
                id="all"
                type="button"
                className="active btn btn-switch btn-bg-light btn-label-sm my-auto mr-2"
                onClick={(e) => this.handleButtonsClick(e)}
              >
                All
              </button>
              <button
                id="InProgress"
                type="button"
                className="btn btn-switch btn-bg-light btn-label-sm my-auto mr-2"
                onClick={(e) => this.handleButtonsClick(e)}
              >
                In Progress
              </button>
              <button
                id="Active"
                type="button"
                className="btn btn-switch btn-bg-light btn-label-sm my-auto mr-2"
                onClick={(e) => this.handleButtonsClick(e)}
              >
                Active
              </button>
              <button
                id="expired"
                type="button"
                className="btn btn-switch btn-bg-light btn-bg-lightbtn-label-sm my-auto mr-2"
                onClick={(e) => this.handleButtonsClick(e)}
              >
                Expired
              </button>
            </div>
            {dataInstances
              .map((dataInstanceClassification) => {
                const instances = dataInstanceClassification.values.map((val) => {
                  const timediff = getTimeCreatedAgo(val.commit.created_at, new Date());
                  const bpipeline = parseToCamelCase(val.backendPipeline);
                  return {
                    id: val.id,
                    currentState: val.status,
                    descTitle: val.name,
                    userName: val.commit.author_name,
                    timeCreatedAgo: timediff,
                    projId: selectedProject.gid,
                    dataOperations: bpipeline.dataOperations,
                    pipelineBackendId: bpipeline.id,
                    inputFiles: bpipeline.inputFiles,
                  };
                });
                const firstValue = dataInstanceClassification.values[0];
                const InstanceName = firstValue && firstValue.name;

                if (instances.length === 0) {
                  return null;
                }
                return (
                  <InstanceCard
                    key={InstanceName}
                    name={InstanceName}
                    history={history}
                    setIsDeleteModalVisible={this.setIsDeleteModalVisible}
                    deletePipeline={this.deletePipeline}
                    setPreconfiguredOPerations={setPreconfiguredOPerations}
                    fireModal={fireModal}
                    params={{
                      currentState: dataInstanceClassification.status,
                      instances,
                    }}
                  />
                );
              })}
          </div>
          <br />
          <br />
        </div>
      </>
    );
  }
}

DataInstanceOverview.propTypes = {
  projects: shape({
    selectedProject: objectOf(shape).isRequired,
  }).isRequired,
  getBranches: func.isRequired,
};

function mapStateToProps(state) {
  return {
    projects: state.projects,
    branches: state.branches,
  };
}

function mapActionsToProps(dispatch) {
  return {
    deleteBranch: bindActionCreators(deleteBranch, dispatch),
    getBranches: bindActionCreators(getBranchesList, dispatch),
    fireModal: bindActionCreators(fireModal, dispatch),
    setPreconfiguredOPerations: bindActionCreators(setPreconfiguredOPerations, dispatch),
  };
}

export default connect(mapStateToProps, mapActionsToProps)(DataInstanceOverview);
