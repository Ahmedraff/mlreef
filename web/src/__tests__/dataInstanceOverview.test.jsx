import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import { BrowserRouter as Router } from 'react-router-dom';
import { storeFactory } from '../functions/testUtils';
import Adapter from 'enzyme-adapter-react-16';
import { DataInstanceOverview } from "../components/data-instance/dataInstanceOverview";
import { projectsArrayMock } from '../testData';
import { Provider } from 'react-redux';

Enzyme.configure({ adapter: new Adapter() });

const store = storeFactory(projectsArrayMock);
const wrapper = shallow(<DataInstanceOverview store={store} />);

describe('Data instance overview contains 4 buttons', () => {
    it('renders filtering functionality buttons', () => {
      expect(wrapper.find('#buttons-container').contains([
        <button>All</button>,
        <button>In Progress</button>,
        <button>Active</button>,
        <button>Expired</button>
      ]));
      expect(wrapper.find('#buttons-container').children()).toHaveLength(4);
      console.log("PASS!!");
    });
  });

  describe('Button filters data instances', () => {
    it('according to the parameter', () => {
      const instanceWrapper = mount((
        <Provider store={store}>
          <Router>
            <DataInstanceOverview />)
          </Router> 
        </Provider>
      ));
      instanceWrapper.find('#buttons-container').childAt(0).simulate('click');
      expect(instanceWrapper).toMatchSnapshot();
    });
  });
  

