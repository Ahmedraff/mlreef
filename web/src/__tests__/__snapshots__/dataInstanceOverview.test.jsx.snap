// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button filters data instances according to the parameter 1`] = `ReactWrapper {}`;
